File: Source/Extensions/Tessa.Extensions.Default.Shared/Workflow/KrProcess/Formatters/NotificationStageTypeFormatter.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Tessa.Platform;
using Tessa.Platform.Storage;
using static Tessa.Extensions.Default.Shared.Workflow.KrProcess.KrConstants;

namespace Tessa.Extensions.Default.Shared.Workflow.KrProcess.Formatters
{
    /// <summary>
    /// Форматтер типа этапа <see cref="StageTypeDescriptors.NotificationDescriptor"/>.
    /// </summary>
    public sealed class NotificationStageTypeFormatter : StageTypeFormatterBase
    {
        #region Base Overrides

        /// <inheritdoc/>
        public override async ValueTask FormatClientAsync(IStageTypeFormatterContext context)
        {
            await base.FormatClientAsync(context);

            var excludeDeputies = context.StageRow.Fields.Get<bool>(KrNotificationSettingVirtual.ExcludeDeputies);
            var excludeSubscribers = context.StageRow.Fields.Get<bool>(KrNotificationSettingVirtual.ExcludeSubscribers);
            var optionalRecipients = context.Card.Sections.TryGet(KrNotificationOptionalRecipientsVirtual.Synthetic)?
                 .Rows
                 .Where(x => x.State != Tessa.Cards.CardRowState.Deleted
                    && x.TryGet<Guid>(KrConstants.StageRowIDReferenceToOwner) == context.StageRow.RowID)
                 .Select(x => x.TryGet<string>(KrNotificationOptionalRecipientsVirtual.RoleName))
                 .ToArray();

            context.DisplayTimeLimit = string.Empty;
            context.DisplaySettings = GetDisplaySettings(excludeDeputies, excludeSubscribers, optionalRecipients);
        }

        /// <inheritdoc/>
        public override async ValueTask FormatServerAsync(IStageTypeFormatterContext context)
        {
            await base.FormatServerAsync(context);

            var excludeDeputies = context.Settings.TryGet<bool>(KrNotificationSettingVirtual.ExcludeDeputies);
            var excludeSubscribers = context.Settings.TryGet<bool>(KrNotificationSettingVirtual.ExcludeSubscribers);
            var optionalRecipients = context.Settings.TryGet<IList>(KrNotificationOptionalRecipientsVirtual.Synthetic)?
                .Cast<Dictionary<string, object>>()
                .Select(x => x.TryGet<string>(KrNotificationOptionalRecipientsVirtual.RoleName))
                .ToArray();

            context.DisplayTimeLimit = string.Empty;
            context.DisplaySettings = GetDisplaySettings(excludeDeputies, excludeSubscribers, optionalRecipients);
        }

        #endregion

        #region Private Methods

        private static string GetDisplaySettings(bool excludeDeputies, bool excludeSubscribers, string[] optionalRecipients)
        {
            var sb = StringBuilderHelper.Acquire();

            if (excludeDeputies)
            {
                sb.Append("{$UI_KrNotification_ExcludeDeputies}");
            }

            if (excludeSubscribers)
            {
                if (sb.Length > 0)
                {
                    sb.AppendLine();
                }

                sb.Append("{$UI_KrNotification_ExcludeSubscribers}");
            }

            if (optionalRecipients is { Length: > 0})
            {
                if (sb.Length > 0)
                {
                    sb.AppendLine();
                }

                sb
                    .Append("{$CardTypes_Controls_OptionalRecipients}: ")
                    .AppendJoin(", ", optionalRecipients);
            }

            return sb.ToStringAndRelease();
        }

        #endregion
    }
}

File: WebClient SDK/src/default/workflow/krProcess/formatters/notificationStageTypeFormatter.ts
import { plainColumnName, sectionName } from 'tessa/workflow';
import { KrStageTypeFormatter, IKrStageTypeFormatterContext, StageTypeHandlerDescriptor,
  notificationDescriptor
} from 'tessa/workflow/krProcess';
import {
  CardRowState
} from 'tessa/cards';

export class NotificationStageTypeFormatter extends KrStageTypeFormatter {

  private _excludeDeputies = plainColumnName('KrNotificationSettingVirtual', 'ExcludeDeputies');
  private _excludeSubscribers = plainColumnName('KrNotificationSettingVirtual', 'ExcludeSubscribers');
  private _optionalRecipients = sectionName('KrNotificationOptionalRecipientsVirtual');

  public descriptors(): StageTypeHandlerDescriptor[] {
    return [notificationDescriptor];
  }

  public format(context: IKrStageTypeFormatterContext) {
    super.format(context);

    const excludeDeputies = context.stageRow.get(this._excludeDeputies);
    const excludeSubscribers = context.stageRow.get(this._excludeSubscribers);
    const optionalRecipients = context.card.sections.tryGet(this._optionalRecipients)!
      .rows
      .filter(x => x.state !== CardRowState.Deleted && x.get("StageRowID") == context.stageRow.rowId)
      .map(x => x.get("RoleName"));

    context.displayTimeLimit = '';
    context.displaySettings = this.getDisplaySettings(excludeDeputies, excludeSubscribers, optionalRecipients);
  }

  private getDisplaySettings(
    excludeDeputies: boolean,
    excludeSubscribers: boolean,
    optionalRecipients: Array<string>
  ): string {
    let settings = '';
    if (excludeDeputies) {
      settings = '{$UI_KrNotification_ExcludeDeputies}';
    }
    if (excludeSubscribers) {
      if (excludeDeputies) {
        settings += '\n';
      }
      settings += '{$UI_KrNotification_ExcludeSubscribers}';
    }
    if (optionalRecipients.length > 0) {
      if (settings.length > 0) {
        settings += '\n';
      }
      settings += '{$CardTypes_Controls_OptionalRecipients}: ';
      optionalRecipients.forEach((x, i) => {
        if (i > 0) {
          settings += ', ';
        }
        settings += x;
      });
    }

    return settings;
  }

}
